sdk/include/ds/xoroshiro.h

```diff
@@ -0,0 +1,75 @@
+// Copyright Microsoft and CHERIoT Contributors.
+// SPDX-License-Identifier: MIT
+//
+// Imported from snmalloc 4f9d991449380ed7d881a25ba02cc5668c1ff394.
+
+#pragma once
+
+#include <cstdint>
+#include <cstdlib>
+
+namespace ds::xoroshiro
+{
+	namespace detail
+	{
+		template<typename STATE, typename RESULT, STATE A, STATE B, STATE C>
+		class XorOshiro
+		{
+			private:
+			static constexpr unsigned StateBits  = 8 * sizeof(STATE);
+			static constexpr unsigned ResultBits = 8 * sizeof(RESULT);
+
+			static_assert(StateBits >= ResultBits,
+			              "STATE must have at least as many bits as RESULT");
+
+			STATE x;
+			STATE y;
+
+			static inline STATE rotl(STATE x, STATE k)
+			{
+				return (x << k) | (x >> (StateBits - k));
+			}
+
+			public:
+			XorOshiro(STATE x = 5489, STATE y = 0) : x(x), y(y)
+			{
+				// If both zero, then this does not work
+				if (x == 0 && y == 0)
+					abort();
+
+				next();
+			}
+
+			void set_state(STATE nx, STATE ny = 0)
+			{
+				// If both zero, then this does not work
+				if (nx == 0 && ny == 0)
+					abort();
+
+				x = nx;
+				y = ny;
+				next();
+			}
+
+			RESULT next()
+			{
+				STATE r = x + y;
+				y ^= x;
+				x = rotl(x, A) ^ y ^ (y << B);
+				y = rotl(y, C);
+				// If both zero, then this does not work
+				if (x == 0 && y == 0)
+					abort();
+				return r >> (StateBits - ResultBits);
+			}
+		};
+	} // namespace detail
+
+	using P128R64 = detail::XorOshiro<uint64_t, uint64_t, 55, 14, 36>;
```

Comment:
What are all of these?  Please give them doc comments.

Reply:
I'll ask @mjp41 where he got his constants. :)